### DISCUSSION TITLE
Up and running. How do I make it permanent?

### DISCUSSION THREAD

**Discussion by kaffeeringe** [owncast/owncast#563] [General Owncast discussion]
Hej!

Thank you for this great project. I used to stream via Twitch. But my stream doesn't benefit from the trashy enviroment there. I get 99% of my viewers through my own invitations. 

It's very easy to download and start owncast. But now it's only running with my terminal open, without https and on port 8080. 

1. How do I turn Owncast into a real service? 
2. How do I do I disable Apache for the webspace to free port 80?
3. Or how do I put my stream through Apache so that I also benefit from the Let's Encrypt certificate there?

**Comment by gabek** [comment]
Hi there, thanks for the kind words!  

Some documentation is actually being worked on to point people in the direction of what they can do after Owncast is up and running in this PR: https://github.com/owncast/owncast.github.io/pull/25/files.  So take a look at that and you should be able to get it running under systemd, as a real service.

Unfortunately we don't have any documentation about using Apache as a reverse proxy, I think most people are using Caddy, Traefik, or nginx.  But it looks like you'll want to do something like this: https://bendellar.com/apache-as-reverse-proxy-for-letsencrypt-free-https-certificates/.

Maybe somebody else can chime in here if they've used Apache as a proxy.

**Reply by kaffeeringe** [reply]
"Apache as a reverse proxy" is at least something I can search for. ;-)

Thank you!

**Reply by geekgonecrazy** [reply]
Make sure you toss websockets in the mix too as the chat portion uses websockets

**Reply by kaffeeringe** [reply]
> Make sure you toss websockets in the mix too as the chat portion uses websockets

Errm. Thank you - now I know why the chat only works on port 8080. But I didn't really find a good help that tells me how to ReverseProxy websockets. Do you have a hint for me?

**Reply by geekgonecrazy** [reply]
Websocket is established at /entry so that’s the one that will be upgraded to websocket connection

**Reply by novadeviator** [reply]
with regard to systemd guide, those steps give me errors: 
```
 owncast.service: Changing to the requested working directory failed: Permission denied
owncast.service: Failed at step CHDIR spawning /root/owncast/owncast: Permission denied
```
the folder with owncast is recursively owned by owncast user.

**Reply by novadeviator** [reply]
I had owncast installation in /root/owncast. I guess non-root users cannot enter /root/ at all. So I moved the folde to /opt/.
I was now getting another error: ```listen tcp 127.0.0.1:81: bind: permission denied```
If I commented the option ```webServerPort: 80``` in config owncast happily started on default port 8080.

**Reply by gabek** [reply]
A non-root user cannot bind to a port below 1024.

**Comment by kaffeeringe** [comment]
I got everything (but the chat yet) to work with Apache as ReverseProxy. My config look a bit like this:

```Apache
<IfModule mod_ssl.c>
        <VirtualHost MYIP:443 >
                ServerName "EXAMPLE.COM"

                SSLEngine on
                SSLVerifyClient none
                SSLCertificateFile /opt/psa/var/certificates/ABC123
                SSLCACertificateFile /opt/psa/var/certificates/ABC123

#Proxy
                SSLProxyEngine On
                ProxyPass / http://localhost:8080/
                ProxyPassReverse / http://localhost:8080/

#extension letsencrypt begin
                Alias /.well-known/acme-challenge "/var/www/MYPATH/.well-known/acme-challenge"

                <Location /.well-known/acme-challenge/>
                        # Require all granted
                        Order Deny,Allow
                        Allow from all
                        Satisfy any
                </Location>

                <LocationMatch "^/.well-known/acme-challenge/(.*/|)\.">
                        # Require all denied
                        Order Allow,Deny
                        Deny from all
                </LocationMatch>
                #extension letsencrypt end
        </VirtualHost>

</IfModule>

<VirtualHost MYIP:80 >
        ServerName "EXAMPLE.COM"
        UseCanonicalName Off

        <IfModule mod_rewrite.c>
                RewriteEngine On
                RewriteCond %{HTTPS} off
                RewriteRule ^ https://%{HTTP_HOST}%{REQUEST_URI} [R=301,L,QSA]
        </IfModule>
</VirtualHost>
```

The main obstacle was, that the owncast user wasn't permitted to access the full path to my owncast installation. I had to chmod 755 /var - maybe the manual should give a suggestion, where to install owncast.

**Reply by nebunez** [reply]
@kaffeeringe I use apache with owncast, here's my config (the websocket part was a copy-paste so I can't explain it):

```  apacheconf
# live.conf
<VirtualHost *:80>
        ServerName live.example.com
        ServerAdmin admin@example.com

        RewriteEngine on
        RewriteCond %{SERVER_NAME} =live.example.com
        RewriteRule ^ https://%{SERVER_NAME}%{REQUEST_URI} [END,NE,R=permanent]
</VirtualHost>

# live-le-ssl.conf
<IfModule mod_ssl.c>
<VirtualHost *:443>
        ServerName live.example.com
        ServerAdmin admin@example.com

        ProxyRequests       Off
        ProxyPreserveHost   On
        AllowEncodedSlashes NoDecode

        <Proxy *>
          Order deny,allow
          Allow from all
        </Proxy>

        ProxyPass        / http://localhost:8080/
        ProxyPassReverse / http://localhost:8080/

        RequestHeader    set X-Forwarded-Proto "https"
        RequestHeader    set X-Forwarded-Port "443"

        # setup the proxy to forward websocket requests properly
        # (note: this proxy automatically converts the secure websocket (wss)
        # to a normal websocket and vice versa, so there's no need to change
        # the colyseus library or the server for that matter)
        RewriteEngine On
        RewriteCond %{HTTP:UPGRADE} ^WebSocket$           [NC,OR]
        RewriteCond %{HTTP:CONNECTION} ^Upgrade$          [NC]
        RewriteRule .* ws://127.0.0.1:8080%{REQUEST_URI}  [P,QSA,L]

        SSLCertificateFile /path/to/fullchain.pem
        SSLCertificateKeyFile /path/to/privkey.pem
        Include /etc/letsencrypt/options-ssl-apache.conf
</VirtualHost>
</IfModule>
```
